### PR TITLE
Overhaul accounts to function correctly

### LINKED ISSUE #50: insufficient funds: 4 but only 4.043, even though settxfee to 0.01
User jdavidof from freenode reports that they received the following err:

`-32603: insufficient funds: transaction requires 4 Coin input but only 4.04366253 Coin spendable`

even after using:

`dcrctl --wallet settxfee 0.01`

they subsequently tried to send 3.99 and it was successfully sent:
https://mainnet.decred.org/tx/ec1a97aef09a6f22d459fc8fd8fdb713613b66b149f946bcae06950968908831

### PR DESCRIPTION
**Pull request by cjepson**
Redesign UTXO storage of accounts in DB, fix accounts
    
Accounts were extremely broken in terms of both spending from them
and finding their balances. This pull request fixes a large number
of issues with them.
    
    - The database now stores the pkScript locations, size, and account
      types. This speeds up pkScript and account pull down immensely,
      and is backwards compatible with the old database format. The
      database version has been bumped to 2.
    - Balances are now calculated for accounts through the wtxmgr's
      database instead of using the old, slow functions.
    - Spending from an individual account has been fixed by refactoring
      the unspentOutputsForAmount function to properly find UTXOs for
      accounts.

Builds on #103, fixes #50, #46, #22, and #2.

### CODE REVIEW

**Comment by jrick** [comment]
I take issue with the way that balances are being returned out of the wallet package.  The previous code would return a struct filled with the total, spendable (by policy) and immature balances.  There may be more fields that should be added there for decred, since that code was added from the latest btcwallet sync.  The advantage of returning balances this way is that calling code that must calculate all of these atomically can do so.  As it is right now, this will break the gRPC server since it will no longer produce correct results all the time.

File: wtxmgr/db.go [hunk 1]
```diff
@@ -70,14 +151,14 @@ var byteOrder = binary.BigEndian
...
-	LatestVersion = 1
+	LatestVersion = 2
 )
 
 // This package makes assumptions that the width of a chainhash.Hash is always 32
 // bytes.  If this is ever changed (unlikely for bitcoin, possible for alts),
 // offsets have to be rewritten.  Use a compile-time assertion that this
 // assumption holds true.
-var _ [32]byte = chainhash.Hash{}
+var _ [chainhash.HashSize]byte = chainhash.Hash{}
```
Comment (jrick) [inline wtxmgr/db.go]: This needs to be reverted.  It used a hardcoded 32 because it needed an assertion that the hash was actually 32 bytes.  This assertion is meaningless.

Comment (cjepson) [inline wtxmgr/db.go]: My mistake, misunderstood the purpose of that.

**Comment by cjepson** [comment]
Looks like I need to add accounts into consolidate finally too, I'll do that now and make it default to the default account.

**Comment by chappjc** [comment]
Trying to test and I get

    src\github.com\decred\dcrwallet\votingpool\withdrawal.go:174: not enough arguments in call to store.AddCredit

Is this an issue for building a functional dcrwallet?

**Comment by cjepson** [comment]
`dcrwallet\votingpool\withdrawal.go:174: not enough arguments in call to store.AddCredit`
Weird, I don't hit that on go install but I'll patch that line anyway. I'm not sure if any of the votingpool stuff works in our wallet.

`2016-03-11 [ERR] WLLT: Notifying unmined tx notification while creating notification for blocks`

This bug was introduced with the merging of the wallet sync branch. jrick is working on patching it.

UTXO consolidation is anti-privacy in the first place, but address reuse won't necessarily fix generation of a large number of UTXOs. Your wallet is splitting UTXOs so it can buy tickets, so anyone stake mining will end up with large numbers of UTXOs in their wallet. The only way around this is periodic calls to 'consolidate'.

File: rpc/rpcserver/server.go [hunk 2]
```diff
@@ -294,6 +294,7 @@ func (s *walletServer) Balance(ctx context.Context, req *pb.BalanceRequest) (
 
 	account := req.AccountNumber
 	reqConfs := req.RequiredConfirmations
+
```
Comment (jrick) [inline rpc/rpcserver/server.go]: Can you remove this extra newline? no reason to touch this file now from this pr.

File: wtxmgr/query.go [hunk 3]
```diff
@@ -517,12 +517,24 @@ func (s *Store) PreviousPkScripts(rec *TxRecord, block *Block) ([][]byte, error)
...
-					if existsRawUnminedCredit(ns, k) == nil {
+					vUC := existsRawUnminedCredit(ns, k)
+					if vUC == nil {
 						continue
 					}
 
-					pkScript, err := fetchRawTxRecordPkScript(
-						prevOut.Hash[:], v, prevOut.Index)
+					// If we encounter an error here, it likely means
+					// we have a legacy outpoint. Ignore the error and
```
Comment (jrick) [inline wtxmgr/query.go]: And if that wasn't the cause of the error?  Have to check this.

File: wtxmgr/query.go [hunk 4]
```diff
@@ -517,12 +517,24 @@ func (s *Store) PreviousPkScripts(rec *TxRecord, block *Block) ([][]byte, error)
...
+					// just let the scrPos be 0, which will trigger
+					// whole transaction deserialization to retrieve
+					// the script.
+					var scrLen uint32
+					scrPos, err := fetchRawUnminedCreditScriptOffset(vUC)
+					if err == nil {
+						scrLen, _ = fetchRawUnminedCreditScriptLength(vUC)
```
Comment (jrick) [inline wtxmgr/query.go]: missed error check

File: wtxmgr/query.go [hunk 5]
```diff
@@ -532,10 +544,22 @@ func (s *Store) PreviousPkScripts(rec *TxRecord, block *Block) ([][]byte, error)
...
+					credVal := existsRawCredit(ns, credKey)
+					if credVal == nil {
+						return storeError(ErrDatabase, fmt.Sprintf("credit "+
+							"val for %x missing", credVal),
+							fmt.Errorf("no exists"))
+					}
+
+					// Legacy outputs in the credit bucket may be of the
+					// wrong size.
+					scrPos, _ := fetchRawCreditScriptOffset(credVal)
```
Comment (jrick) [inline wtxmgr/query.go]: missed error check

File: wtxmgr/query.go [hunk 6]
```diff
@@ -532,10 +544,22 @@ func (s *Store) PreviousPkScripts(rec *TxRecord, block *Block) ([][]byte, error)
...
+					scrLen, _ := fetchRawCreditScriptLength(credVal)
```
Comment (jrick) [inline wtxmgr/query.go]: missed error check

File: wtxmgr/query.go [hunk 7]
```diff
@@ -550,9 +574,23 @@ func (s *Store) PreviousPkScripts(rec *TxRecord, block *Block) ([][]byte, error)
...
+			credVal := existsRawCredit(ns, credKey)
+			if credVal == nil {
+				return storeError(ErrDatabase, fmt.Sprintf("credit "+
+					"val for %x missing in debit itr", credVal),
+					fmt.Errorf("no exists"))
+			}
+
+			// Legacy credit output values may be of the wrong
+			// size.
+			scrPos, _ := fetchRawCreditScriptOffset(credVal)
```
Comment (jrick) [inline wtxmgr/query.go]: missed error check

File: wtxmgr/query.go [hunk 8]
```diff
@@ -550,9 +574,23 @@ func (s *Store) PreviousPkScripts(rec *TxRecord, block *Block) ([][]byte, error)
...
+			scrLen, _ := fetchRawCreditScriptLength(credVal)
```
Comment (jrick) [inline wtxmgr/query.go]: missed error check

File: wtxmgr/tx.go [hunk 9]
```diff
@@ -574,6 +576,75 @@ func (s *Store) PruneUnconfirmed(height int32, stakeDiff int64) error {
 	return nil
 }
 
+// fetchAccountForPkScript fetches an account for a given pkScript given a
+// credit value, the script, and an account lookup function. It does this
+// to maintain compatibility with older versions of the database.
+func (s *Store) fetchAccountForPkScript(credVal []byte, unminedCredVal []byte,
+	pkScript []byte,
+	acctLookupFunc func(dcrutil.Address) (uint32, error)) (uint32, error) {
```
Comment (jrick) [inline wtxmgr/tx.go]: acctLookupFunc is already a Store field.

File: wtxmgr/tx.go [hunk 10]
```diff
@@ -574,6 +576,75 @@ func (s *Store) PruneUnconfirmed(height int32, stakeDiff int64) error {
...
+			}
+			switch storeErr.Code {
+			case ErrNoExists:
+			case ErrData:
+			default:
+				return 0, err
+			}
+		}
+
+		if err == nil {
```
Comment (jrick) [inline wtxmgr/tx.go]: For readibility, I'd move the err==nil case to the top right after the function call, and since this diverges (returns from the function), the error handling code can be unindented

File: wtxmgr/tx.go [hunk 11]
```diff
@@ -574,6 +576,75 @@ func (s *Store) PruneUnconfirmed(height int32, stakeDiff int64) error {
...
+			return acct, nil
```
Comment (jrick) [inline wtxmgr/tx.go]: Same here

File: wtxmgr/tx.go [hunk 12]
```diff
@@ -574,6 +576,75 @@ func (s *Store) PruneUnconfirmed(height int32, stakeDiff int64) error {
...
+			pkScript, s.chainParams)
+	if err != nil {
+		return 0, err
+	}
+
+	// Only look at the first address returned. This does not
+	// handle multisignature or other custom pkScripts in the
+	// correct way, which requires multiple account tracking.
+	acct, err := acctLookupFunc(addrs[0])
+	if err != nil {
```
Comment (jrick) [inline wtxmgr/tx.go]: error check here is unneeded, you can just return the result of the function call.

File: wtxmgr/tx.go [hunk 13]
```diff
@@ -681,11 +752,30 @@ func (s *Store) moveMinedTx(ns walletdb.Bucket, rec *TxRecord, recKey,
 		cred.opCode = fetchRawUnminedCreditTagOpcode(it.cv)
 		cred.isCoinbase = fetchRawUnminedCreditTagIsCoinbase(it.cv)
 
+		// Legacy credit output values may be of the wrong
+		// size.
+		scrType, _ := fetchRawUnminedCreditScriptType(it.cv)
+		scrPos, _ := fetchRawUnminedCreditScriptOffset(it.cv)
+		scrLen, _ := fetchRawUnminedCreditScriptLength(it.cv)
```
Comment (jrick) [inline wtxmgr/tx.go]: 3 more missed error checks...

File: wtxmgr/tx.go [hunk 14]
```diff
@@ -898,8 +988,46 @@ func getP2PKHOpCode(pkScript []byte) uint8 {
 	return OP_NONSTAKE
 }
 
+// pkScriptType determines the general type of pkScript for the purposes of
+// fast extraction of pkScript data from a raw transaction record.
+func pkScriptType(pkScript []byte) (scriptType, error) {
+	var scrType scriptType
```
Comment (jrick) [inline wtxmgr/tx.go]: I would just return inside the switch statement, so that the switch statement becomes a terminating statement (needs a default case too).  If a new script class is added and not all cases are handled, this will return the zero value with no error, which is probably wrong.

File: wtxmgr/query.go (hunk 3 above)
Comment (cjepson) [inline wtxmgr/query.go]: No, it states in that comment why it can be skipped (a zero value scrPos/scrLen is handled appropiately).

File: wtxmgr/query.go (hunk 4 above)
Comment (cjepson) [inline wtxmgr/query.go]: Same as above

File: wtxmgr/query.go (hunk 3 above)
Comment (cjepson) [inline wtxmgr/query.go]: Oh, I see you, mean the second error. If the first one succeeds the second one should not fail. I can add error handling there.

File: wtxmgr/query.go (hunk 5 above)
Comment (cjepson) [inline wtxmgr/query.go]: These are intentional (see comment, scrPos/scrLen being zero is fine).

File: wtxmgr/query.go (hunk 6 above)
Comment (cjepson) [inline wtxmgr/query.go]: See above

File: wtxmgr/tx.go (hunk 9 above)
Comment (cjepson) [inline wtxmgr/tx.go]: Thanks, forgot to remove this.

File: wtxmgr/tx.go [hunk 15]
```diff
@@ -1542,6 +1702,133 @@ func (s *Store) UnspentOutputs() ([]*Credit, error) {
...
+	if minedCredV == nil && unminedCredV == nil {
+		errStr := fmt.Errorf("missing utxo %x, %v", op.Hash, op.Index)
+		return nil, storeError(ErrNoExists, "couldn't find relevant credit "+
+			"for unspent output", errStr)
+	}
+
+	// Throw an inconsistency error if we find one.
+	if minedCredV != nil && unminedCredV != nil {
+		errStr := fmt.Errorf("duplicated utxo %x, %v", op.Hash, op.Index)
+		return nil, storeError(ErrNoExists, "credit exists in mined and unmined "+
```
Comment (jrick) [inline wtxmgr/tx.go]: I don't think this is what ErrNoExists means.

File: wtxmgr/tx.go (hunk 12 above)
Comment (cjepson) [inline wtxmgr/tx.go]: I consider this stylistic, I strongly prefer to end all function with declared nil errors rather than returning a function.

File: wtxmgr/tx.go (hunk 15 above)
Comment (cjepson) [inline wtxmgr/tx.go]: Yeah, I was using this incorrectly. I'll add a new error specific to Vs missing in the database.

File: wtxmgr/db.go [hunk 16]
```diff
@@ -538,6 +619,41 @@ func readRawTxRecordBlock(k []byte, block *Block) error {
 	return nil
 }
 
+func extractRawTxRecordPkScript(k, v []byte, index uint32, scrLoc uint32,
```
Comment (jrick) [inline wtxmgr/db.go]: I don't think this should have been renamed.  As commented above, extract slices an existing slice to extract a value (still memory mapped!) while fetch returns an entirely new value or allocation.

File: wtxmgr/db.go [hunk 17]
```diff
@@ -538,6 +619,41 @@ func readRawTxRecordBlock(k []byte, block *Block) error {
...
+		scrLocInt := int(scrLoc) + int64Size
+		scrLenInt := int(scrLen)
+		pkScript = make([]byte, scrLenInt)
+		copy(pkScript, v[scrLocInt:scrLocInt+scrLenInt])
+	}
+
+	return pkScript, nil
+}
+
+func extractRawTxRecordReceived(v []byte) time.Time {
```
Comment (jrick) [inline wtxmgr/db.go]: fetch

**Comment by jrick** [comment]
So I've looked over all of the new database code, and here are my honest thoughts.

I am pretty displeased that database changes being made here.  This is a lot of code for a temporary solution to a problem that may not be a problem to begin with (we need correct account behavior, not *faster* account behavior).  If the code was unit tested I would feel more comfortable with this change (while unit tests can not prove correctness, they are still useful to catch regressions), however, none of the wtxmgr unit tests were updated for decred (or even compile) and all of these changes will need to be tested on live systems.  This is very brittle and I believe it will be the cause of many issues down the road.

Additionally, due to the nature of this being a temporary fix, and one that I will not be making in btcwallet, I expect issues in a future code sync after btcwallet correctly integrates the address (key) and transaction managers into a single namespace.  Even if bugs are avoided in this change, it's very likely that others will be introduced in that sync.

I can not in good conscious give this change an OK, but I have looked through the code and pointed out the most glaring of issues, of which most have been corrected (and the others just haven't had time to be fixed yet).  Hopefully that feedback is useful.

**Comment by cjepson** [comment]
I'm not terribly pleased with this code either and I may fix and test it more. Despite this, we do need working accounts in our wallet that are reasonably fast for users with large wallets. I think this code is the best we can do without completely ripping out waddrmgr and wtxmgr and replacing them with something with the desired efficiency and scalability. The update runs without issue on the simulation network with both old and new wallets. It's the best I can do for now and the users need something with the correct functionality in a fast time frame that won't fail on large datasets.

**Comment by jrick** [comment]
Regarding testing on live systems, I'm currently running this branch with Paymetheus and I'm not seeing any issues.

**Comment by alexlyp** [comment]
tACK, did extensive simnet testing with account name usage.  appears to fix all previous issues with sendfrom and accounts
